Fixed #30360 -- Support rotation of secret keys.

carltongibson opened the pull request:
Take 2 on #13618. 

@apollo13 New PR, as per your suggestion. This one passes the tests at least. It's has edges, but if you could give it your thoughts, that would be amazing. 

As per your thoughts: Normalise in `Settings.__init__()` and raise if empty in `LazySettings.__getattr__()`. 

I needed to normalise in `override_settings.__init__()`, because `enable()` was too hard™. 
What are you thoughts here? Does it look OK? 

Had to add a `secret` property in PasswordTokenGenerator to handle lazy reading the secret value from SECRET_KEYS. 

Do you think it's viable. Happy to work it if so. 
Clearly it needs docs. 
What test do you want to see? 

Thanks! 🚀

sumanthratna commented on django/conf/__init__.py:
```diff
@@ -86,8 +86,17 @@ def __getattr__(self, name):
+        elif name == 'SECRET_KEYS':
+            if not val:
+                raise ImproperlyConfigured('The SECRET_KEYS setting must not be empty.')
+            if not val[0]:
```
nit: since the error message says "the elements" shouldn't this check all elements?
```suggestion
            if not all(element for element in val):
```
alternatively, we can leave the if-statement as-is and instead change the error to something like "the first key in the list must not be empty"

carltongibson replied:
The system check already does that. This is just a runtime safety net really. We can extend it or adjust the message but I'm not sure we need to handle it too robustly.

apollo13 commented on django/contrib/auth/tokens.py:
```diff
@@ -12,20 +15,42 @@ class PasswordResetTokenGenerator:
+    # SECRET_KEYS.
+    @property
+    def secret(self):
+        return self._secret or self._get_secrets()[0]
```
It is not exactly clear to me why this would need to be a property instead of calling doing:
```
self.secret = self.secret or self._getsecrets()[0]
```
I see that you might get into some troubles in `_get_secret` then. But this begs the question: We do not document the `secret` attribute at all (at least I didn't see it), so we could just remove that in favor of `secrets`. If we want to be nice a deprecation path could look like this:
```
def __init__(self):
  if self.secret:
    self.secrets = [self.secret]
    warnings.warn("Using deprecated feature set, blablabla")
  self.secrets = self.secrets or settings.SECRET_KEYS
```
Then you can simply drop the properties and `get_secrets` all together.

apollo13 commented on django/contrib/auth/tokens.py:
```diff
@@ -47,14 +72,21 @@ def check_token(self, user, token):
+            constant_time_compare(self._make_token_with_timestamp(user, ts, secret), token)
+            for secret in self._get_secrets()
+        ]
+        if not constant_time_any(attempts):
```
Is this really necessary? Which information do we leak by using a normal any? To the best of my knowledge you'd leak which `SECRET_KEY` you used, which imo shouldn't be a problem? I also think it is maybe not the best idea to try against all secret keys if the first one (the common one) already matches, this creates extra work that should not be needed imo.

In that sense I'd loop and break out early if you get a valid sig.

pelme replied:
This was not fully clear when I wrote the original patch either. 🙂 I added constant_time_any just for good measure. I discussed it with @MarkusH at DjangoCon EU in Copenhagen and (IIRC) we figured that it would be a good idea to keep the constant time check just to be extra careful.

That said, yes, it could leak which secret key that was used but... I cannot see why that would be a problem. I do not have a strong opinion either way.

apollo13 commented on django/core/signing.py:
```diff
@@ -72,10 +74,16 @@ def base64_hmac(salt, value, key, algorithm='sha1'):
 def get_cookie_signer(salt='django.core.signing.get_cookie_signer'):
     Signer = import_string(settings.SIGNING_BACKEND)
-    key = force_bytes(settings.SECRET_KEY)  # SECRET_KEY may be str or bytes.
-    return Signer(b'django.http.cookies' + key, salt=salt)
```
While I realize we cannot change that now, do we remember why we added `django.http.cookies` here? The salt alone should make sure that we do not clash with other signatures.

carltongibson replied:
> ...do we remember why... 

No we don't 😄 
It was [part of the initial implementation a decade ago](https://github.com/django/django/commit/f60d42846365b2bf2f1c9bc7a3007c303122a20b#diff-855893e869078af0cac503844c1a016958ffa925614d70de68340c3cd52f6e55R87).

MarkusH replied:
If the user is able to make the server sign arbitrary data of their choosing, using the same salt and key will yield the same signature. However, if we slightly modify the key, they can't. This makes chosen plaintext attacks less vulnerable.

apollo13 reviewed (changes requested):
>  What test do you want to see?

I didn't look through them just yet to give you an unbiased answer. At the minimum I'd like to see tests that:

 * Verify the system check works with a few weird combos of the secret keys
 * Tests where we verify that the second (and an even later) key in the list is actually used to decode
 * Tests where we verify that the first key is actually used for signing and not just any key :D
 
 The rest should be covered by our existing tests I think.

carltongibson commented:
Hey @apollo13 — Thanks for the review. I've made various adjustments/simplifications based on your comments. 

* Adjusted the SECRET_KEY(S) normalisation in `Settings.__init__()`. (Happy for you to suggest a different approach there...) 
* Check `all()` keys in `LazySettings.__getattr__()`
* Removed the `constant_time_any()`
* Moved `PasswordResetTokenGenerator` directly to `secrets` (It's not public API but I will add release note for this last as a breaking change.)

Can I get you to take another look? 

I will make the docs adjustments in the meantime. 

Thanks 👍

carltongibson commented:
OK, just an update, we're going to target this for 4.0. It makes sense at the high-level but:

* As nice as it would be, it's too close to the feature freeze — It's too hot™: it's a sensitive area and it needs a slower review. 
* There are a whole load of overlapping changes here (including #12553 which is still to come in) such that ^^^ is amplified. 
* We're dropping all pre-Django 3.1 token support in 4.0, which means a whole another batch of similarly overlapping changes. 
* and so on... 

Bumping to 4.0 takes the pressure off of getting all of that right (in too tight a time frame). We've lasted 15 years without key rotation. 8 more months will be fine.

apollo13 commented:
> Can I get you to take another look?

Sure, given that we are pushing this out of 3.2, I'll take a little bit more time with it. Remind me if I forget :)

apollo13 commented:
@carltongibson I lost track of that, what is the current status here? Really like to have that in 4.0

kezabelle commented on django/conf/__init__.py:
```diff
@@ -185,6 +194,19 @@ def __init__(self, settings_module):
+            self.SECRET_KEYS = [self.SECRET_KEY]
+        elif self.SECRET_KEYS and not self.SECRET_KEY:
+            self.SECRET_KEY = self.SECRET_KEYS[0]
+
```
In case it's relevant, this patches up normal usages of settings - that is, using a settings module and `Settings` - but leaves `UserSettingsHolder` without the same shoring up (although I may be mis-reading the bigger picture, apologies if so). 
That is, if I do `django.conf.settings.configure(SECRET_KEY='...')` I won't end up with a `SECRET_KEYS` where `[0] == SECRET_KEY`, I think? Less of a problem going forward (users will just add the key), only an issue for extant usages getting upgraded.
It'll still error with `ImproperlyConfigured` correctly on accessing the lazy setting, mind. 
I know some things are intentionally left to the caller (`tzset` shenanigans are omitted). Is this one of those?

carltongibson replied:
Hey @kezabelle. Thanks for the comments. Thinking about both now. 

This one: 

> ...leaves UserSettingsHolder without the same shoring up...

I'm thinking that a mapping in `__getattr__` might be the way to go — _We missed the one lookup, let's try the other_ — I think we probably need that for compatibility around the ecosystem. ... 🤔 — I'll add a test case and see.

kezabelle commented on django/core/checks/security/base.py:
```diff
@@ -190,17 +190,22 @@ def check_ssl_redirect(app_configs, **kwargs):
-            not secret_key.startswith(SECRET_KEY_INSECURE_PREFIX)
-        )
+        passed_check = all(check_key(key) for key in secret_keys)
+
```
Is it worth checking that `secret_keys` != `isinstance(str)` (and raise a separate warning if it is)? It seems easy enough to accidentally do `SECRET_KEYS='a...b'` and whilst you'll still get the deployment warning (because each character won't pass, naturally) it won't _necessarily_ be obvious why, if your setting looks like it's otherwise a 'good' candidate. Moreso if you consider that suffixing your accidental string usage, with a comma would fix it (turning it into a tuple) or vice versa.

MarkusH replied:
Good idea!

carltongibson replied:
I think adding `SECRET_KEYS` to `tuple_settings` covers this. See e5300963294a62b7321aa7e99f802e4d8557dd80

carltongibson commented:
Hey @apollo13 — I've updated this, sketched out the docs adjustments, and addressed previous comments. Can I get you to take a look again? 

I'm looking at @kezabelle's comments now. (Handle `UserSettingsHolder`, and tighten up the system checks.) 

We're essentially using `SECRET_KEYS[0]` internally. So the bit I'm 😬 about is the mapping back to `SECRET_KEY`, that I think we need, since everything out there is using it. (…)

ngnpope commented on docs/ref/settings.txt:
```diff
@@ -2258,13 +2258,46 @@ The secret key is used for:
 If you rotate your secret key, all of the above will be invalidated.
 Secret keys are not used for passwords of users and key rotation will not
-affect them.
+affect them. Use :setting:`SECRET_KEYS` to allow secret key rotation.
```
I get that we don't want to deprecate `SECRET_KEY` at this stage, but do we want to add an admonition that encourages use of `SECRET_KEYS` more strongly now? This would, hopefully, make it easier to do a more strict deprecation later, if that decision is made, as we've highlighted the preferred setting as early as possible to encourage wider use.

apollo13 commented on docs/ref/settings.txt:
```diff
@@ -2258,13 +2258,46 @@ The secret key is used for:
+    ]
+
+… new values will be signed with ``'new-secret-key'``, whereas checks will be
+made against both ``'new-secret-key'`` and ``'new-secret-key'``.
```
should say `old-secret-key` in the second occurance

apollo13 commented on django/core/signing.py:
```diff
@@ -166,7 +180,10 @@ def unsign(self, signed_value):
         value, sig = signed_value.rsplit(self.sep, 1)
-        if constant_time_compare(sig, self.signature(value)):
+        attempts = [
+            constant_time_compare(sig, self.signature(value, key)) for key in self.keys
```
We are doing a lot of work here when usually only the first key should be needed. I understand that usually the list will for short; but still…

what about:
```
for key in self.keys:
  valid = constant_time_compare(sig, self.signature(value, key))
  if valid:
    return value
```

ngnpope replied:
Is there any risk inherent in having multiple comparisons though?

We call `constant_time_compare()` intentionally to reduce the risk of timing attacks. If it were possible to trigger something to be unsigned against `SECRET_KEYS[0]` (one key used) and something against `SECRET_KEYS[1]` (two keys used), does that pose a problem? If anything we ought to *always* compare against all keys. Perhaps we should add a comment here to ensure that happens and a warning in the docs for `SECRET_KEYS` that all keys need to be checkd, so don't add too many?

apollo13 replied:
The "risk" is that it would add a constant overhead to every signing operation if there is more than one entry in the list. I currently cannot imagine an attack where knowing which of the secret keys (index number) was used -- but I'll happily be proven wrong.

Adding a doc comment would somewhat alleviate my concern since it would make it clear that `keys[1:]` is basically just transitional and should be removed asap. That said it simply feels wrong to perform unneeded work -- I know that this code is security sensitive and I will probably regret that I said that this work is unneeded when one indeed finds a security issue in the future… but still :)

ngnpope replied:
> The "risk" is that it would add a constant overhead to every signing operation if there is more than one entry in the list. I currently cannot imagine an attack where knowing which of the secret keys (index number) was used -- but I'll happily be proven wrong.

Me neither, but thought it worth raising for consideration.

> Adding a doc comment would somewhat alleviate my concern since it would make it clear that `keys[1:]` is basically just transitional and should be removed asap.

+1 to that.

> That said it simply feels wrong to perform unneeded work

Oh, I'd be the first to agree with that.

> I know that this code is security sensitive and I will probably regret that I said that this work is unneeded when one indeed finds a security issue in the future… but still :)

😅

kezabelle replied:
Fair warning, I don't know nearly enough about things of this nature to be actually valuable, except possibly to ask questions which might be dumb or might be previously unconsidered. So please take it with a grain of salt...

If an adversary is making use of the fact a secret key has been compromised, which scenario gives them less information when the key is subsequently discovered and rotated and they want to stop trying to make use of it (to fly under the radar as much as possible); one constant time compare, or multiple? Layman's gut feeling says you'd want the time comparison to be actually roughly constant, which would probably require always doing either 1 or 2 (I think, it's murky in my brain now) always, regardless of number of keys in the corpus? (ie: if there's only one key, always compare as if there are 2...)

And does that kind of scenario _actually_ matter? IDK.

apollo13 commented on django/contrib/auth/tokens.py:
```diff
@@ -50,7 +50,11 @@ def check_token(self, user, token):
 
         # Check that the timestamp/uid has not been tampered with
-        if not constant_time_compare(self._make_token_with_timestamp(user, ts), token):
+        attempts = [
```
Same as below, I think I'd prefer a loop that breaks early.

carltongibson replied:
I've gone with this: 

```diff
         # Check that the timestamp/uid has not been tampered with
-        attempts = [
-            constant_time_compare(self._make_token_with_timestamp(user, ts, secret), token)
-            for secret in self.secrets
-        ]
-        if not any(attempts):
+        valid_compare = False
+        for secret in self.secrets:
+            valid_compare = constant_time_compare(self._make_token_with_timestamp(user, ts, secret), token)
+            if valid_compare:
+                break
+        if not valid_compare:
             return False
```

(Incoming). Happy to take style points on that. 😄 (But I think it's what you meant)

apollo13 replied:
What about:
```py
for secret in self.secrets:
    if constant_time_compare(self._make_token_with_timestamp(user, ts, secret), token):
        break
    else:
        return False        
```

carltongibson replied:
@florian. Yes! That's nicer. (incoming.) 👍

carltongibson commented:
Hi @kezabelle, @ngnpope, @apollo13 — thanks for taking an interest in this one! (Important to get it right.)

@ngnpope and @apollo13 — I accept your points above — +1 — I'll work those in. 

I was thinking about @kezabelle's review comments above… 

Specifically, the thing that's been bothering me is trying to keep the values of `SECRET_KEY` and `SECRET_KEYS` in sync. It's two sources of truth, and that's (pretty much) always an error. 

Thinking about this for `UserSettingsHolder`, I thought the best way would be if `SECRET_KEY` became a property, aliasing to the (now) one source of truth. 

I added that in bc0a4482eb86e3a2494083abdc378ec484d67166. Can I ask you to take a look? What do you think? Everything passes, and it removed the duplicate `Normalize SECRET_KEY(S)` kludges. 

If we think that's neater I can clean up, document `SECRET_KEYS` as preferred,  and that `SECRET_KEYS[1:]` incur a penalty (so shouldn't hang around), adjusting the signing checks as suggested, and so on.  

Thanks!

felixxm commented on django/conf/__init__.py:
```diff
@@ -280,5 +310,20 @@ def __repr__(self):
             'cls': self.__class__.__name__,
         }
 
+    @property
+    def SECRET_KEY(self):
+        try:
+            return self.SECRET_KEYS[0]
+        except TypeError:
+            return ''
+
+    @SECRET_KEY.setter
+    def SECRET_KEY(self, value):
+        self.SECRET_KEYS = [value]
+
+    @SECRET_KEY.deleter
+    def SECRET_KEY(self):
+        del self.SECRET_KEYS
+
```
Do we need setters/getters and extra code in `is_overridden()`? Maybe it enough to set `SECRET_KEYS` in `Settings.__init__()`, e.g.

```python
class Settings:
    def __init__(self, settings_module):
        ...
        if self.is_overridden('SECRET_KEY'):
            if self.is_overridden('SECRET_KEYS'):
                raise ImproperlyConfigured(
                    'SECRET_KEY and SECRET_KEYS are mutually exclusive.'
                )
            setattr(self, 'SECRET_KEYS', [self.SECRET_KEY])
```
I would add `SECRET_KEYS` to the `tuple_settings`.

carltongibson replied:
> Do we need setters/getters ... ?

So this is the question in https://github.com/django/django/pull/13850/commits/2f1d3a397e621a1a0f82a91bbd00e7c950599dc6 — does it look better (is it more robust) having one source of truth for the two settings, vs trying to keep them in sync? (Could go either way here, but pushed this to get opinions.) 

The property removes the need to do the same (flaky?) dance in `override_settings` — and note `UserSettingsHolder` wasn't covered at all before. 

> ... and extra code in is_overridden()

That was just the minimum I needed to get the tests to pass. Maybe/likely there's a better way of handling that.

MarkusH replied:
I'm leaning towards @felixxm idea of having a single point of truth and not needing to keep them in sync. Keeping something in sync is always hard and error prone. But if there's only one way to get to `settings.SECRET_KEY` we can encapsulate all logic in there.

carltongibson replied:
Hmmm. This was (≈) the original idea, but it was difficult to handle all the cases: providing SECRET_KEY, providing SECRET_KEYS, making sure the values were sane when mapping between the options (in `Settings`, in `UserSettingsHolder`, in `override_settings()`) — it was precisely the issue of keeping two sources of truth in sync. 

That's why I went for the `@property` approach. `SECRET_KEY` defers then (always) to `SECRET_KEYS`. `override_settings` is then not altered at all (which is a win.) I thought the diff in 2f1d3a3 made it much cleaner, as well as having it work for `UserSettingsHolder` along with `Settings`. 

There **is** a bit of duplication currently between `Settings` and `UserSettingsHolder` that we **could** move into a mixin. (As ever, we could go either way on what's cleaner there. 🤷‍♂️)

This is the main bit to pin down with @felixxm next week.

carltongibson commented:
Hi all — thanks again for the input here! 

* I've pulled in @apollo13's suggestion ref breaking from the comparison loops. 
* Adjusted settings docs to point strongly towards the new SECRET_KEYS, given that SECRET_KEY is now a (mere) proxy, that we **might** remove some day. 

I've asked @felixxm to give it his fully-considered look next week, but any other comments very much appreciated. Are we going to do this? Are we happy with the general approach? 

* I will go over the checks wording and other occurrences of `SECRET_KEY` this afternoon. 
* I'll leave the multiple commits for the moment to leave the context for review.

carltongibson commented:
OK, I've gone through all the docs and tests to use `SECRET_KEYS` (unless specifically testing `SECRET_KEY`). 

The one thing on my list is to looks at the usages in the checks (and checks.txt) which I'll do next week. 

Then it's _Any other comments_ and getting it past @felixxm 😄. 

I think it's a nice feature, but if there are objection, please do raise 'em now. 😝

MarkusH commented on django/conf/global_settings.py:
```diff
@@ -266,11 +266,16 @@ def gettext_noop(s):
 SECRET_KEY = ''
 
+# A list of secret keys used in hashing algorithms for this particular Django
+# installation. The first key in the list will be used for signatures. The
+# other keys will be used to verify the validity of signatures to allow key
+# rotation. Set this in your settings, or Django will complain loudly.
```
This makes it sound like the second to nth key are the only one used for validation, when in fact all keys are used for validation.

carltongibson commented:
Hey @felixxm. I've squashed this and rebased. 

* Below is the diff moving to the property. The original way was trying to sync the two settings values, but I think the property comes out much cleaner. (There's still the duplication with `UserSettingsHolder` which could be reduced.) Thoughts welcome! 
* Current docs learn to _"Use `SECRET_KEYS`!"_ but `SECRET_KEY` could be fine for most folks, so we could adjust the messaging there to "Use `SECRET_KEYS` if you need key rotation."
* What extra tests do you want? 😄 


<details>
<summary>diff moving to property, rather than trying to keep values in-sync</summary>

```diff

From 44947f9c44e28dc4f20cea6546584e87ab016740 Mon Sep 17 00:00:00 2001
From: Carlton Gibson <carlton.gibson@noumenal.es>
Date: Thu, 11 Nov 2021 10:49:19 +0100
Subject: [PATCH] Made SECRET_KEY a property on Settings/UserSettingsHolder.

---
 django/conf/__init__.py         | 46 ++++++++++++++++++++++++++-------
 django/test/utils.py            |  8 ------
 tests/auth_tests/test_tokens.py |  2 +-
 3 files changed, 37 insertions(+), 19 deletions(-)

diff --git a/django/conf/__init__.py b/django/conf/__init__.py
index dfd7235477..279c359c4e 100644
--- a/django/conf/__init__.py
+++ b/django/conf/__init__.py
@@ -186,6 +186,9 @@ class Settings:
         self.SETTINGS_MODULE = settings_module
 
         mod = importlib.import_module(self.SETTINGS_MODULE)
+        if hasattr(mod, 'SECRET_KEY') and hasattr(mod, 'SECRET_KEYS'):
+            if mod.SECRET_KEY and mod.SECRET_KEYS:
+                raise ImproperlyConfigured('Only one of SECRET_KEY and SECRET_KEYS can be specified, not both.')
 
         tuple_settings = (
             'ALLOWED_HOSTS',
@@ -202,6 +205,8 @@ class Settings:
                         not isinstance(setting_value, (list, tuple))):
                     raise ImproperlyConfigured("The %s setting must be a list or a tuple." % setting)
                 setattr(self, setting, setting_value)
+                if setting == 'SECRET_KEY':
+                    setting = 'SECRET_KEYS'
                 self._explicit_settings.add(setting)
 
         if self.USE_TZ is False and not self.is_overridden('USE_TZ'):
@@ -212,16 +217,6 @@ class Settings:
                 category=RemovedInDjango50Warning,
             )
 
-        # Normalize SECRET_KEY/SECRET_KEYS when either is provided.
-        if not (self.SECRET_KEY or self.SECRET_KEYS):
-            pass  # Neither is allowed in script usage.
-        elif self.SECRET_KEY and self.SECRET_KEYS:
-            raise ImproperlyConfigured('Only one of SECRET_KEY and SECRET_KEYS can be specified, not both.')
-        elif self.SECRET_KEY:
-            self.SECRET_KEYS = [self.SECRET_KEY]
-        else:
-            self.SECRET_KEY = self.SECRET_KEYS[0]
-
         if self.is_overridden('USE_DEPRECATED_PYTZ'):
             warnings.warn(USE_DEPRECATED_PYTZ_DEPRECATED_MSG, RemovedInDjango50Warning)
 
@@ -241,6 +236,8 @@ class Settings:
             warnings.warn(USE_L10N_DEPRECATED_MSG, RemovedInDjango50Warning)
 
     def is_overridden(self, setting):
+        if setting == 'SECRET_KEY':
+            setting = 'SECRET_KEYS'
         return setting in self._explicit_settings
 
     def __repr__(self):
@@ -249,6 +246,18 @@ class Settings:
             'settings_module': self.SETTINGS_MODULE,
         }
 
+    @property
+    def SECRET_KEY(self):
+        return self.SECRET_KEYS[0]
+
+    @SECRET_KEY.setter
+    def SECRET_KEY(self, value):
+        self.SECRET_KEYS = [value]
+
+    @SECRET_KEY.deleter
+    def SECRET_KEY(self):
+        del self.SECRET_KEYS
+
 
 class UserSettingsHolder:
     """Holder for user configured settings."""
@@ -289,6 +298,8 @@ class UserSettingsHolder:
         )
 
     def is_overridden(self, setting):
+        if setting == 'SECRET_KEY':
+            setting = 'SECRET_KEYS'
         deleted = (setting in self._deleted)
         set_locally = (setting in self.__dict__)
         set_on_default = getattr(self.default_settings, 'is_overridden', lambda s: False)(setting)
@@ -299,5 +310,20 @@ class UserSettingsHolder:
             'cls': self.__class__.__name__,
         }
 
+    @property
+    def SECRET_KEY(self):
+        try:
+            return self.SECRET_KEYS[0]
+        except TypeError:
+            return ''
+
+    @SECRET_KEY.setter
+    def SECRET_KEY(self, value):
+        self.SECRET_KEYS = [value]
+
+    @SECRET_KEY.deleter
+    def SECRET_KEY(self):
+        del self.SECRET_KEYS
+
 
 settings = LazySettings()
diff --git a/django/test/utils.py b/django/test/utils.py
index cb9672c91d..81005d4f8f 100644
--- a/django/test/utils.py
+++ b/django/test/utils.py
@@ -455,14 +455,6 @@ class override_settings(TestContextDecorator):
     enable_exception = None
 
     def __init__(self, **kwargs):
-        # Normalize SECRET_KEY(S)
-        has_secret_key = 'SECRET_KEY' in kwargs
-        has_secret_keys = 'SECRET_KEYS' in kwargs
-        if has_secret_key and not has_secret_keys:
-            kwargs['SECRET_KEYS'] = [kwargs['SECRET_KEY']] if kwargs['SECRET_KEY'] else []
-        if has_secret_keys and not has_secret_key:
-            kwargs['SECRET_KEY'] = kwargs['SECRET_KEYS'][0] if kwargs['SECRET_KEYS'] else None
-
         self.options = kwargs
         super().__init__()
 
diff --git a/tests/auth_tests/test_tokens.py b/tests/auth_tests/test_tokens.py
index 5f9ee69a75..cf39cb7b93 100644
--- a/tests/auth_tests/test_tokens.py
+++ b/tests/auth_tests/test_tokens.py
@@ -136,7 +136,7 @@ class TokenGeneratorTest(TestCase):
     @override_settings(SECRET_KEY='')
     def test_secret_lazy_validation(self):
         default_token_generator = PasswordResetTokenGenerator()
-        msg = 'The SECRET_KEYS setting must not be empty.'
+        msg = 'The elements of SECRET_KEYS must not be empty.'
         with self.assertRaisesMessage(ImproperlyConfigured, msg):
             default_token_generator.secrets
 
-- 
2.31.1


```

</details>

ngnpope commented:
> The original way was trying to sync the two settings values, but I think the property comes out much cleaner.

Perhaps clearer and more robust in some ways. Not sure about cleaner given the need for the following in three places:

```diff
+        if setting == 'SECRET_KEY':
+            setting = 'SECRET_KEYS'
```

Are we sure we have nailed this in all the right places? Could this impact on third-party apps that do things with settings?

> (There's still the duplication with `UserSettingsHolder` which could be reduced.) Thoughts welcome!

The property stuff is nearly identical except for the catching of `TypeError`. It could be moved into a `SecretKeySettingCompatMixin` 😅 and then use a class variable to determine whether to suppress the exception...

I guess if it were possible to make the mixin more meaty, de-duplicating more of the stuff in `Settings` and `UserSettingsHolder`, the issue of the `if settings == SECRET_KEY:` duplication may also be reduced.

carltongibson commented:
> …and then use a class variable to determine whether to suppress the exception...

That difference — catching the `TypeError` —  is a point to review: I caught the exception in the one-place to get the tests to pass — but it may be that it rightly belongs in both. 🤔

carltongibson commented:
Quick update: 

From a suggestion coming from discussion with @felixxm and @apollo13 — we're going to try a slightly different approach, keeping `SECRET_KEY` as the main value and adding a `SECRET_KEY_FALLBACKS`, which would default to an empty list, to allow folks to do the rotation here. 

The `SECRET_KEYS` proposal is fine but 99% of folks would just have a single value, never rotate, and likely not get fully why the list was needed. (There's also a semantic point about the list head being significant.) 

Sticking with `SECRET_KEY` keeps that clearer. Adding the `SECRET_KEY_FALLBACKS` should allow clearer messaging around _If you want to rotate_, and means we avoid this _mapping_ difficulty, that's proved, erm..., difficult. 

I'll sketch that out next.

pelme commented:
That is a lot better than overloading/deprecating/replacing SECRET_KEY with SECRET_KEYS. And fully backwards compatible!

pegler commented:
Thanks for your efforts on this pull! I'm curious if this still has a chance of making it into 4.0?

felixxm commented:
> Thanks for your efforts on this pull! I'm curious if this still has a chance of making it into 4.0?

Unfortunately no, the Django 4.0 feature freeze was on September 20, 2021.

tim-schilling commented:
@carltongibson I took a swing at implementing the fallback keys that you mentioned. The branch is at https://github.com/tim-schilling/django/tree/ticket-30360-fallback-keys. I did make an opinionated approach on supplying a new parameter, `read_keys` to `Signer`. It makes sense to me to identify the keys that can only be used for validation (perhaps I should have named it `validator_keys`). I also kept the same pattern regarding the setters and getters in `PasswordResetTokenGenerator`.

carltongibson commented:
Hey @tim-schilling — super! Do you want to open a PR there? 🥇

carltongibson commented:
Closing in favour of @tim-schilling's new PR. Thanks all for the input so far.
